### ISSUE TITLE
compress/zlib.reader returns unexpected EOF when calling Read() after EOF

### ISSUE THREAD

**Issue by AsamK** [golang/go#14675]
1. What version of Go are you using (`go version`)?
go1.6

2. What operating system and processor architecture are you using (`go env`)?
linux/amd64

3. What did you do?
Test returning of EOF and unexpected EOF
https://play.golang.org/p/RJfp3HgcQ5

4. What did you expect to see?
The last read should return EOF

5. What did you see instead?
The last read returns UnexpectedEOF
The issue was triggered by b1797390b95d1ffd3d97b19532bf451719d42fd5 , which introduced unexpected EOF for truncated streams.

I'm not certain, if the io.Read() interface guarantees that calls after the first one that returned EOF should also return EOF, but returning "unexpected EOF" is definitely unexpected.

**Comment by AsamK** [comment]
A simple fix would be to always store the error returned by z.decompressor.read:
```diff
diff --git a/src/compress/zlib/reader.go b/src/compress/zlib/reader.go
index 78ea704..640089f 100644
--- a/src/compress/zlib/reader.go
+++ b/src/compress/zlib/reader.go
@@ -94,8 +94,10 @@ func (z *reader) Read(p []byte) (n int, err error) {
 
 	n, err = z.decompressor.Read(p)
 	z.digest.Write(p[0:n])
-	if n != 0 || err != io.EOF {
+	if (err != nil) {
 		z.err = err
+	}
+	if n != 0 || err != io.EOF {
 		return
 	}
 
```

**Comment by bradfitz** [comment]
> I'm not certain, if the io.Read() interface guarantees that calls after the first one that returned EOF should also return EOF, but returning "unexpected EOF" is definitely unexpected.

The io.Reader interfaces guarantees nothing about any calls past the first error, but I suppose this is fixable, especially if behavior changes from Go 1.5 to Go 1.6.

**Comment by bradfitz** [comment]
```
bradfitz@laptop ~$ GOROOT=$HOME/go1.4 go run zread.go
go1.4.2
4 err: <nil>
0 err: EOF
0 err: EOF
bradfitz@laptop ~$ GOROOT=$HOME/go1.5 go run zread.go
go1.5.1
4 err: <nil>
0 err: EOF
0 err: EOF
bradfitz@laptop ~$ GOROOT=$HOME/go1.6 go run zread.go
go1.6
4 err: <nil>
0 err: EOF
0 err: unexpected EOF
```

**Comment by dsnet** [comment]
I'll fix this. There's another unrelated bug here. The zlib package doesn't reset z.err when Reset is called.

Adding the following to @AsamK's example still reports io.ErrUnexpectedEOF:
```go
reader.(zlib.Resetter).Reset(bytes.NewReader(compress.Bytes()), nil)
fmt.Println(reader.Read(buf[:])) // 0 unexpected EOF
```

**Comment by AsamK** [comment]
Thanks for fixing this so quickly, will the fix be included in a 1.6.1 release?

**Comment by bradfitz** [comment]
No promises, but we'll consider it.

**Comment by bradfitz** [comment]
@rsc, @ianlancetaylor, @adg, this was a regression from Go 1.4 and Go 1.5 and presumably earlier. Maybe a candidate for Go 1.6.1.

Related is #14676

**Comment by rsc** [comment]
Not sure this is critical enough for 1.6.1. Is it breaking something important that can't be changed not to read multiple times at EOF or to use a fixed copy of the library?

**Comment by dsnet** [comment]
Also, @bradfitz's fix in #14676 could easily be applied by the user as well.

**Comment by rsc** [comment]
This is unfortunate but a big change. The mime instance was the main trigger and we have a simpler mime fix for 1.6.2.
